Shop_Unity_base/Assets/Scripts/Shop/Model/ShopModel.cs

```diff
@@ -2,166 +2,182 @@
         private void AddMessage(string message)
         {
-            messages.Add(message);
-            while (messages.Count > MaxMessageQueueCount)
+            messages.Enqueue(message);
+
+            while (messages.Count > MAX_MESSAGES_IN_QUEUE)
+                messages.Dequeue();
+        }
+
+        public void SetCustomer(Customer customer)
+        {
+            this.customer = customer;
+        }
+
+        public Item Buy()
+        {
+            Item item = GetSelectedItem();
+
+            if (customer.Gold < item.Cost)
+            {
+                AddMessage("You do not have enough gold to buy that item!");
+                AddMessage("Item cost: " + item.Cost + ", Your gold: " + customer.Gold);
+                return null;
+            }
+
+            OnBuyEvent?.Invoke(item, customer);
+            return item;
+        }
+
+        // We assume that the customer has the same type of items that the shop has
+        public void Sell()
+        {
+            Item item = GetSelectedItem();
+
+            if (gold >= 0 && gold < item.Cost)
             {
-                messages.RemoveAt(0);
+                AddMessage("The shop does not have enough gold to buy that item!");
+                return;
             }
+
+            OnSellEvent?.Invoke(item, customer);
         }
 
-        //------------------------------------------------------------------------------------------------------------------------
-        //                                                  Buy()
-        //------------------------------------------------------------------------------------------------------------------------        
-        //not implemented yet
-        public void Buy()
+        // TODO sent messages have old data
+        private void OnCustomerBuyItem(Item item, Customer customer)
         {
-            AddMessage("You can't buy this item yet!");
+            AddGold(item.Cost);
+            itemList.Remove(item);
+
+            AddMessage($"{customer.Name} purchased [{item.name}] from the shop for -{item.Cost} gold");
+            AddMessage($"{customer.Name}'s gold: {customer.Gold}");
         }
 
-        //------------------------------------------------------------------------------------------------------------------------
-        //                                                  Sell()
-        //------------------------------------------------------------------------------------------------------------------------        
-        //not implemented yet
-        public void Sell()
+        private void OnCustomerSellItem(Item item, Customer customer)
         {
-            AddMessage("You can't sell this item yet!");
+            AddGold(-item.Cost);
+            itemList.Add(item);
+
+            AddMessage($"{customer.Name} sold [{item.name}] to the shop for +{item.Cost} gold");
+            AddMessage($"{customer.Name}'s gold: {customer.Gold}");
         }
```

Comment:
Since you already pass the customer as part of the delegate, it would make sense if the Shop is the one responsible for updating the Customer. However, I do understand the intention of this.

If you must do it this way, I would suggest removing the customer requirement from the delegate. Then you will have to remove the string messages too, but at that point, you could move the messaging system out of the Shop (or Models in general) and have it be its own (static) object and have multiple Models communicate with it, such as the Customer (to retain the original functionality). An example would be:

```csharp
// In ShopModel
private void OnCustomerBuyItem(Item item)
{
//...
AddMessage($"The shop sold [{item.name} for {item.Cost} gold. Gold: {Gold}");
}

// In Customer
private void OnBuyItem(Item item)
{
//...
AddMessage($"Customer {name} bought [{item.name}] for {item.Cost} gold. Gold: {Gold}"); 
}

Reply:
I like it.